instrumentation/grpc-1.40.0/src/main/java/com/nr/agent/instrumentation/grpc/GrpcUtil.java

```diff
@@ -0,0 +1,63 @@
+/*
+ *
+ *  * Copyright 2022 New Relic Corporation. All rights reserved.
+ *  * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package com.nr.agent.instrumentation.grpc;
+
+import com.newrelic.agent.bridge.Token;
+import com.newrelic.api.agent.NewRelic;
+import com.newrelic.api.agent.Transaction;
+import io.grpc.Metadata;
+import io.grpc.Status;
+
+public class GrpcUtil {
+
+    /**
+     * Finalize the transaction when a Stream closed or is cancelled by linking the supplied token,
+     * setting the appropriate headers and marking the transaction as sent.
+     *
+     * @param token The {@link Token} to expire
+     * @param status The {@link Status} of the completed/cancelled operation
+     * @param metadata Operation {@link Metadata} to be included with the transaction
+     */
+    public static void finalizeTransaction(Token token, Status status, Metadata metadata) {
+        if (token != null) {
+            token.link();
+            Transaction transaction = NewRelic.getAgent().getTransaction();
+            transaction.setWebResponse(new GrpcResponse(status, metadata));
+            transaction.addOutboundResponseHeaders();
+            transaction.markResponseSent();
+        }
+    }
+
+    /**
+     * Set the response.status attribute and error cause (if applicable) on the transaction
+     * when the ServerStream is closed or cancelled.
+     *
+     * @param status The {@link Status} of the completed/cancelled operation
+     */
+    public static void setServerStreamResponseStatus(Status status) {
+        if (status != null) {
+            NewRelic.addCustomParameter("response.status", status.getCode().value());
+            if (GrpcConfig.errorsEnabled && status.getCause() != null) {
+                // If an error occurred during the close of this server call we should record it
+                NewRelic.noticeError(status.getCause());
+            }
+        }
+    }
+
+    /**
+     * Expire the supplied {@link Token} instance if it's non-null.
+     *
+     * @param token The {@link Token} to expire
+     */
+    public static void expireToken(Token token) {
+        if (token != null) {
+            token.expire();
+            token = null;
```

Comment:
I don't think this method is very useful.
The token that gets nulled out here is the reference that is used in this method.
If while debugging, you step out of this method, you'll see that the reference that was passed to this method is still referencing the original.

Reply:
Good catch ... I'll refactor and remove this method